substrate/frame/meta-tx/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+pub enum Proof<Address, Signature> {
+	Signed(Address, Signature),
+	// TODO `General` as in `sp_runtime::generic::Preamble`.
+}
+
+pub type MetaTxFor<T> = MetaTx<
+	<<T as Config>::PublicKey as IdentifyAccount>::AccountId,
+	<T as Config>::Signature,
+	<T as Config>::RuntimeCall,
+	<T as Config>::TxExtension,
+>;
+
+pub type SignedPayloadFor<T> =
+	sp_runtime::generic::SignedPayload<<T as Config>::RuntimeCall, <T as Config>::TxExtension>;
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The overarching call type.
+		type RuntimeCall: Parameter
+			+ GetDispatchInfo
+			+ Dispatchable<
+				Info = DispatchInfo,
+				PostInfo = PostDispatchInfo,
+				RuntimeOrigin = Self::RuntimeOrigin,
+			> + IsType<<Self as frame_system::Config>::RuntimeCall>;
+		/// Signature type for meta transactions.
+		type Signature: Parameter + Verify<Signer = Self::PublicKey>;
+		/// Public key type used for signature verification.
+		type PublicKey: IdentifyAccount<AccountId = Self::AccountId>;
+		/// The context type of `Self::TxExtension`
+		type TxContext: Member + Default;
+		/// Transaction extension/s for meta transactions.
+		type TxExtension: TransactionExtension<<Self as Config>::RuntimeCall, Self::TxContext>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// TODO
+		TODO,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A call was dispatched.
+		Dispatched { result: DispatchResultWithPostInfo },
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Dispatch meta transaction.
+		///
+		/// origin must be signed
+		#[pallet::call_index(0)]
+		#[pallet::weight({
+			let dispatch_info = meta_tx.call.get_dispatch_info();
+			// TODO: plus T::WeightInfo::dispatch() which must include the weight of T::TxExtension
+			(
+				dispatch_info.weight,
+				dispatch_info.class,
+			)
+		})]
+		pub fn dispatch(origin: OriginFor<T>, meta_tx: MetaTxFor<T>) -> DispatchResultWithPostInfo {
+			let _who = ensure_signed(origin)?;
+
+			let (signer, signature) = match meta_tx.proof {
+				Proof::Signed(signer, signature) => (signer, signature),
+			};
+
+			let signed_payload = SignedPayloadFor::<T>::new(*meta_tx.call, meta_tx.tx_ext)
+				.map_err(|_| Error::<T>::TODO)?;
+
+			if !signed_payload.using_encoded(|payload| signature.verify(payload, &signer)) {
+				return Err(Error::<T>::TODO.into());
+			}
+
+			let origin = T::RuntimeOrigin::signed(signer);
+			let (call, tx_ext, _) = signed_payload.deconstruct();
+			let info = call.get_dispatch_info();
+			// TODO: to get the len we have to encode the original `meta_tx`.
+			let len = 0;
+			let mut ctx = T::TxContext::default();
+
```

Comment:
we can use set the `DispatchTransaction` bound for `Config::TxExtension` and use it's impl for everything below. here, for the demonstration purpose, I'm not utilizing it.

Reply:
`DispatchTransaction` is meant to work only for "legacy" extensions which use a `()` context. I believe the current impl is the right thing :+1: .